2003-01-21  Anton Kachalov <mouse@altlinux.org>

	* apt-pkg/versionmatch.cc (pkgVersionMatch::Find):
	Add APT::Install::VirtualVersion support.
	* cmdline/apt-get.cc (TryToChangeVer):
	Substitute virtual package with real one.

--- apt-0.5.4cnc9/apt-pkg/versionmatch.cc.orig	2002-07-23 21:54:50 +0400
+++ apt-0.5.4cnc9/apt-pkg/versionmatch.cc	2003-01-21 16:42:55 +0300
@@ -15,6 +15,7 @@
 #pragma implementation "apt-pkg/versionmatch.h"
 #endif
 #include <apt-pkg/versionmatch.h>
+#include <apt-pkg/configuration.h>
 
 #include <apt-pkg/strutl.h>
 #include <apt-pkg/error.h>
@@ -158,6 +159,19 @@ pkgCache::VerIterator pkgVersionMatch::Find
 	    return Ver;
    }
       
+   if (_config->FindB("APT::Install::VirtualVersion",false) == false)
+      return Ver;
+
+   pkgCache::PrvIterator Prv = Pkg.ProvidesList();
+   for (; Prv.end() == false; Prv++)
+    {
+      if (Type == Version)
+      {
+	 if (MatchVer(Prv.ProvideVersion(),VerStr,VerPrefixMatch) == true)
+	    return Prv.OwnerVer();
+	 continue;
+      }
+    }
    // This will be Ended by now.
    return Ver;
 }
--- apt-0.5.4cnc9/cmdline/apt-get.cc.orig	2002-10-18 22:32:12 +0400
+++ apt-0.5.4cnc9/cmdline/apt-get.cc	2003-01-23 15:33:45 +0300
@@ -1119,7 +1119,7 @@
 // TryToChangeVer - Try to change a candidate version			/*{{{*/
 // ---------------------------------------------------------------------
 /* */
-bool TryToChangeVer(pkgCache::PkgIterator Pkg,pkgDepCache &Cache,
+bool TryToChangeVer(pkgCache::PkgIterator &Pkg,pkgDepCache &Cache,
 		    const char *VerTag,bool IsRel)
 {
    pkgVersionMatch Match(VerTag,(IsRel == true?pkgVersionMatch::Release : 
@@ -1143,6 +1143,9 @@
    }
    
    Cache.SetCandidateVersion(Ver);
+   if (Ver.ParentPkg() != Pkg) {
+      Pkg = Ver.ParentPkg();
+   }
    return true;
 }
 									/*}}}*/
